federation/pkg/federation-controller/namespace/namespace_controller.go

```diff
@@ -0,0 +1,353 @@
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Core().Namespaces().Watch(options)
+			},
+		},
+		&api.Namespace{},
+		controller.NoResyncPeriodFunc(),
+		framework.ResourceEventHandlerFuncs{
+			// Trigger reconcilation whenever some defiition is changed.
+			DeleteFunc: func(old interface{}) { nc.deliverNamespaceObj(old, 0, 0) },
+			AddFunc:    func(cur interface{}) { nc.deliverNamespaceObj(cur, 0, 0) },
+			UpdateFunc: func(old, cur interface{}) {
+				nold, _ := old.(*api_v1.Namespace)
+				ncur, _ := old.(*api_v1.Namespace)
+				// Object meta or spec updated. Status changes are ignored.
+				if !reflect.DeepEqual(nold.Spec, ncur.Spec) ||
+					!reflect.DeepEqual(nold.ObjectMeta, ncur.ObjectMeta) {
+					nc.deliverNamespaceObj(cur, 0, 0)
+				}
+			},
+		},
+	)
+
+	// Federated informer on namespaces in members of federation.
+	nc.namespaceFederatedInformer = util.NewFederatedInformer(
+		client,
+		func(cluster *federation_api.Cluster, targetClient federation_release_1_4.Interface) (cache.Store, framework.ControllerInterface) {
+			return framework.NewInformer(
+				&cache.ListWatch{
+					ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+						return targetClient.Core().Namespaces().List(options)
+					},
+					WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+						return targetClient.Core().Namespaces().Watch(options)
+					},
+				},
+				&api.Namespace{},
+				controller.NoResyncPeriodFunc(),
+				framework.ResourceEventHandlerFuncs{
+					// Trigger reconcilation whenever something in federated cluster is changed. In most cases it
+					// would be just confirmation that some namespace opration suceeded.
+					DeleteFunc: func(old interface{}) {
+						// Remove once it is done automatically.
+						nold, _ := old.(*api_v1.Namespace)
+						util.SetClusterName(nold, cluster.Name)
+						nc.deliverNamespaceObj(old, NamespaceReviewDelay, 0)
+					},
+					AddFunc: func(cur interface{}) {
+						// Remove once it is done automatically.
+						ncur, _ := cur.(*api_v1.Namespace)
+						util.SetClusterName(ncur, cluster.Name)
+						nc.deliverNamespaceObj(cur, NamespaceReviewDelay, 0)
+					},
+					UpdateFunc: func(old, cur interface{}) {
+						nold, _ := old.(*api_v1.Namespace)
+						ncur, _ := cur.(*api_v1.Namespace)
+						// Remove once it is done automatically.
+						util.SetClusterName(nold, cluster.Name)
+						util.SetClusterName(ncur, cluster.Name)
+
+						if !reflect.DeepEqual(nold.Spec, ncur.Spec) ||
+							!reflect.DeepEqual(nold.ObjectMeta, ncur.ObjectMeta) {
+							nc.deliverNamespaceObj(cur, NamespaceReviewDelay, 0)
+						}
+					},
+				})
+		},
+		&util.ClusterLifecycleHandlerFuncs{
+			ClusterAvailable: func(cluster *federation_api.Cluster) {
+				// When new cluster becomes available process all the namespaces again.
+				nc.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(ClusterAvailableDelay))
```

Comment:
In every place you are using deliverer, the pattern is:
```
...DeliverAt(.., time.Now().Add(delay))
```

What not just having a method:
```
DeliverAfter()
```
(e.g.: ...DeliverAfter(..., delay) - it would be simpler).